File: client2/src/ecs/System.ts
import { Aspect } from './Aspect'
import { Entity } from './Entity'
import { Manager } from './Manager'

export class System {
  static _id = 'System'
  _id = System._id

  static _aspect = new Aspect()
  _aspect = System._aspect

  debug: boolean = false
  entities: Map<number, Entity>
  manager: Manager
  paused: boolean

  constructor(manager: Manager) {
    this.entities = new Map()
    this.manager = manager
    this.paused = false
  }

  setDebug = (debug: boolean) => {
    this.debug = debug
  }

  log = (...args: any[]) => {
    this.debug && print(...args)
  }

  add = (entity: Entity) => {
    this.entities.set(entity.id, entity)
  }

  check = (entity: Entity) => {
    if (this._aspect.check(entity)) {
      if (!this.entities.get(entity.id)) {
        this.onAdd(entity)

        this.log(
          `entity (id :${entity.id}) is now being processed by system (id: ${
            this._id
          })`
        )
      }

      this.entities.set(entity.id, entity)
    } else {
      if (this.entities.get(entity.id)) {
        this.onRemove(entity)
        this.log(
          `entity (id :${
            entity.id
          }) is no longer being processed by system (id: ${this._id})`
        )
      }

      this.entities.delete(entity.id)
    }
  }

  keyboard = (
    key: string,
    scancode: Scancode,
    isRepeat: boolean,
    isPressed: boolean
  ) => {}

  mouse = (
    x: number,
    y: number,
    button: number,
    isTouch: boolean,
    isPressed: boolean
  ) => {}

  pause = () => {
    this.paused = true
  }

  remove = (entity: Entity) => {
    this.entities.delete(entity.id)
  }

  resume = () => {
    this.paused = false
  }

  beginContact = (a: Fixture, b: Fixture, contact: Contact) => {}

  endContact = (a: Fixture, b: Fixture, contact: Contact) => {}

  draw = () => {}

  onAdd = (entity: Entity) => {}

  onRemove = (entity: Entity) => {}

  update = (dt: number) => {}
}

File: client/src/ecs/system.lua
local Aspect = require 'src.ecs.aspect'

local System = {}

function System:new(id, aspect)
  local system = {
    id = id,
    aspect = aspect or Aspect.new(),
    entities = {},
    manager = nil,
    paused = false
  }

  setmetatable(system, self)
  self.__index = self

  return system
end

function System:setManager(manager)
  self.manager = manager
end

function System:check(entity)
  if (self.aspect:check(entity)) then
    if not self.entities[entity.id] then
      self:onAdd(entity)
    end

    self:add(entity)
  else
    if self.entities[entity.id] then
      self:onRemove(entity)
    end

    self:remove(entity)
  end
end

function System:add(entity)
  self.entities[entity.id] = entity
end

function System:onAdd(entity)
end

function System:remove(entity)
  self.entities[entity.id] = nil
end

function System:onRemove(entity)
end

function System:pause()
  self.paused = true
end

function System:resume()
  self.paused = false
end

function System:keyboard(key, scancode, isrepeat, ispressed)
end

function System:mouse(x, y, button, istouch, presses)
end

function System:beginContact(a, b, contact)
end

function System:endContact(a, b, contact)
end

function System:update(dt)
end

function System:draw()
end

return System
